File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        pair = {
            "(": ")",
            "{": "}",
            "[": "]",
        }
        stack = []
        
        for c in s:
            if c in pair:
                stack.append(c)
            else:
                if not stack:
                    return False
                
                prev = stack.pop()
                if c != pair[prev]:
                    return False
        
        return len(stack) == 0

File: 0020-valid-parentheses/0020-valid-parentheses.go
func isValid(s string) bool {
	p := []rune{}

	for _, j := range s {
		switch j {
		case '[', '(', '{':
			p = append(p, j)

		case ']':
			if len(p) == 0 || p[len(p)-1] != '[' {
				return false
			}
			p = p[:len(p)-1]

		case ')':
			if len(p) == 0 || p[len(p)-1] != '(' {
				return false
			}
			p = p[:len(p)-1]
		case '}':
			if len(p) == 0 || p[len(p)-1] != '{' {
				return false
			}
			p = p[:len(p)-1]
		}
	}

	return len(p) == 0
}